rasa/core/channels/botframework.py

```diff
@@ -188,25 +201,93 @@ def __init__(self, app_id: Text, app_password: Text) -> None:
         self.app_id = app_id
         self.app_password = app_password
 
+        self.jwt_keys: Dict[Text, Any] = {}
+        self.jwt_update_time = datetime.datetime.fromtimestamp(0)
+
+        self._update_cached_jwk_keys()
+
+    def _update_cached_jwk_keys(self) -> None:
+        logger.debug("Updating JWT keys for the Botframework.")
+        response = requests.get(MICROSOFT_OPEN_ID_URI)
+        response.raise_for_status()
+        conf = response.json()
+
+        jwks_uri = conf["jwks_uri"]
+
+        keys_request = requests.get(jwks_uri)
+        keys_request.raise_for_status()
+        keys_list = keys_request.json()
+        self.jwt_keys = {key["kid"]: key for key in keys_list["keys"]}
+        self.jwt_update_time = datetime.datetime.now()
+
+    def _validate_jwt_token(self, jwt_token: Text) -> None:
+        jwt_header = jwt.get_unverified_header(jwt_token)  # type: ignore
+        key_id = jwt_header["kid"]
+        if key_id not in self.jwt_keys:
+            raise InvalidKeyError(f"JWT Key with ID {key_id} not found.")
+
+        key_json = self.jwt_keys[key_id]
+        public_key = RSAAlgorithm.from_jwk(key_json)  # type: ignore
+        jwt.decode(
+            jwt_token,
+            key=public_key,
+            audience=self.app_id,
+            algorithms=jwt_header["alg"],
+        )
+
+    def _validate_auth(self, auth_header: Optional[Text]) -> Optional[HTTPResponse]:
+        if not auth_header:
+            return response.text(
+                "No authorization header provided", status=HTTPStatus.UNAUTHORIZED
+            )
+
+        # Update the JWT keys daily
+        if datetime.datetime.now() - self.jwt_update_time > datetime.timedelta(days=1):
+            try:
+                self._update_cached_jwk_keys()
+            except HTTPError as error:
+                logger.warning(
+                    f"Could not update JWT keys from {MICROSOFT_OPEN_ID_URI}."
+                )
+                logger.exception(error, exc_info=True)
+
+        auth_match = BEARER_REGEX.match(auth_header)
+        if not auth_match:
+            return response.text(
+                "No Bearer token provided in Authorization header",
+                status=HTTPStatus.UNAUTHORIZED,
```

Comment:
Should this be a `BadRequest` instead of `Unauthorized`?

Reply:
I don't think so, see above.